File: app/src/main/java/com/dougkeen/bart/data/FollowedTripRecord.java
package com.dougkeen.bart.data;

import com.dougkeen.bart.model.Departure;
import com.dougkeen.bart.model.Line;
import com.dougkeen.bart.model.Station;
import com.dougkeen.bart.model.TripLeg;
import com.dougkeen.bart.model.TripStop;

import java.util.ArrayList;
import java.util.List;

/** Versioned JSON schema for the durable followed-trip state. */
public final class FollowedTripRecord {
    public static final int CURRENT_VERSION = 1;

    public int version = CURRENT_VERSION;
    public String origin;
    public String trainDestination;
    public String passengerDestination;
    public String line;
    public String trainDestinationColorHex;
    public String trainDestinationColorText;
    public String platform;
    public String direction;
    public boolean bikeAllowed;
    public String trainLength;
    public boolean requiresTransfer;
    public boolean transferScheduled;
    public boolean limited;
    public boolean canceled;
    public boolean listedInETDs = true;
    public int minutes;
    public long minEstimate;
    public long maxEstimate;
    public long arrivalTimeOverride;
    public int estimatedTripTime;
    public List<TripLegRecord> tripLegs = new ArrayList<>();

    public FollowedTripRecord() {
    }

    public static FollowedTripRecord fromDeparture(Departure departure) {
        FollowedTripRecord record = new FollowedTripRecord();
        record.origin = abbreviation(departure.getOrigin());
        record.trainDestination = abbreviation(departure.getTrainDestination());
        record.passengerDestination = abbreviation(departure.getPassengerDestination());
        record.line = departure.getLine() == null
                ? null : departure.getLine().name();
        record.trainDestinationColorHex = departure.getTrainDestinationColorHex();
        record.trainDestinationColorText = departure.getTrainDestinationColorText();
        record.platform = departure.getPlatform();
        record.direction = departure.getDirection();
        record.bikeAllowed = departure.isBikeAllowed();
        record.trainLength = departure.getTrainLength();
        record.requiresTransfer = departure.getRequiresTransfer();
        record.transferScheduled = departure.isTransferScheduled();
        record.limited = departure.isLimited();
        record.canceled = departure.isCanceled();
        record.listedInETDs = departure.isListedInETDs();
        record.minutes = departure.getMinutes();
        record.minEstimate = departure.getMinEstimate();
        record.maxEstimate = departure.getMaxEstimate();
        record.arrivalTimeOverride = departure.getArrivalTimeOverride();
        record.estimatedTripTime = departure.getEstimatedTripTime();
        for (TripLeg leg : departure.getTripLegs()) {
            record.tripLegs.add(TripLegRecord.fromTripLeg(leg));
        }
        return record;
    }

    public Departure toDeparture() {
        if (version != CURRENT_VERSION) {
            throw new IllegalArgumentException(
                    "Unsupported followed trip format version: " + version);
        }
        Departure departure = new Departure();
        departure.setOrigin(station(origin));
        departure.setTrainDestination(station(trainDestination));
        departure.setPassengerDestination(station(passengerDestination));
        departure.setLine(line == null ? null : Line.valueOf(line));
        departure.setTrainDestinationColorHex(trainDestinationColorHex);
        departure.setTrainDestinationColorText(trainDestinationColorText);
        departure.setPlatform(platform);
        departure.setDirection(direction);
        departure.setBikeAllowed(bikeAllowed);
        departure.setTrainLength(trainLength);
        departure.setRequiresTransfer(requiresTransfer);
        departure.setTransferScheduled(transferScheduled);
        departure.setLimited(limited);
        departure.setCanceled(canceled);
        departure.setListedInETDs(listedInETDs);
        departure.setMinutes(minutes);
        departure.setMinEstimate(minEstimate);
        departure.setMaxEstimate(maxEstimate);
        departure.setArrivalTimeOverride(arrivalTimeOverride);
        departure.setEstimatedTripTime(estimatedTripTime);
        List<TripLeg> legs = new ArrayList<>();
        if (tripLegs != null) {
            for (TripLegRecord leg : tripLegs) {
                legs.add(leg.toTripLeg());
            }
        }
        departure.setTripLegs(legs);
        return departure;
    }

    private static String abbreviation(Station station) {
        return station == null ? null : station.abbreviation;
    }

    private static Station station(String abbreviation) {
        return abbreviation == null ? null
                : Station.getByAbbreviation(abbreviation);
    }

    public static final class TripLegRecord {
        public String line;
        public String origin;
        public String destination;
        public String trainDestination;
        public String tripId;
        public long departureTime;
        public long arrivalTime;
        public List<TripStopRecord> stops = new ArrayList<>();

        public TripLegRecord() {
        }

        static TripLegRecord fromTripLeg(TripLeg leg) {
            TripLegRecord record = new TripLegRecord();
            record.line = leg.getLine() == null ? null : leg.getLine().name();
            record.origin = abbreviation(leg.getOrigin());
            record.destination = abbreviation(leg.getDestination());
            record.trainDestination = abbreviation(leg.getTrainDestination());
            record.tripId = leg.getTripId();
            record.departureTime = leg.getDepartureTime();
            record.arrivalTime = leg.getArrivalTime();
            for (TripStop stop : leg.getStops()) {
                record.stops.add(TripStopRecord.fromTripStop(stop));
            }
            return record;
        }

        TripLeg toTripLeg() {
            TripLeg leg = new TripLeg();
            leg.setLine(line == null ? null : Line.valueOf(line));
            leg.setOrigin(station(origin));
            leg.setDestination(station(destination));
            leg.setTrainDestination(station(trainDestination));
            leg.setTripId(tripId);
            leg.setDepartureTime(departureTime);
            leg.setArrivalTime(arrivalTime);
            List<TripStop> tripStops = new ArrayList<>();
            if (stops != null) {
                for (TripStopRecord stop : stops) {
                    tripStops.add(stop.toTripStop());
                }
            }
            leg.setStops(tripStops);
            return leg;
        }
    }

    public static final class TripStopRecord {
        public String station;
        public long arrivalTime;
        public long departureTime;

        public TripStopRecord() {
        }

        static TripStopRecord fromTripStop(TripStop stop) {
            TripStopRecord record = new TripStopRecord();
            record.station = abbreviation(stop.getStation());
            record.arrivalTime = stop.getArrivalTime();
            record.departureTime = stop.getDepartureTime();
            return record;
        }

        TripStop toTripStop() {
            return new TripStop(station(station), arrivalTime, departureTime);
        }
    }
}

File: app/src/main/kotlin/in/izyum/bart/data/FollowedTripRecord.kt
package `in`.izyum.bart.data

import `in`.izyum.bart.model.Itinerary
import `in`.izyum.bart.model.Line
import `in`.izyum.bart.model.PredictionSource
import `in`.izyum.bart.model.Station
import `in`.izyum.bart.model.TripLeg
import `in`.izyum.bart.model.TripStop
import java.time.LocalDate

/** Versioned JSON schema for the durable followed-trip state. */
class FollowedTripRecord {
    @JvmField var version: Int = CURRENT_VERSION
    @JvmField var origin: String? = null
    @JvmField var trainDestination: String? = null
    @JvmField var passengerDestination: String? = null
    @JvmField var line: String? = null
    @JvmField var platform: String? = null
    @JvmField var trainLength: String? = null
    @JvmField var canceled = false
    @JvmField var minutes = 0
    @JvmField var minEstimate = 0L
    @JvmField var maxEstimate = 0L
    @JvmField var arrivalTimeOverride = 0L
    @JvmField var estimatedTripTime = 0
    @JvmField var tripLegs: MutableList<TripLegRecord> = mutableListOf()

    fun toItinerary(): Itinerary {
        require(version == CURRENT_VERSION) {
            "Unsupported followed trip format version: $version"
        }
        val itineraryOrigin = station(origin)
            ?: throw IllegalArgumentException("Followed trip has no origin")
        val itineraryDestination = station(passengerDestination ?: trainDestination)
            ?: throw IllegalArgumentException("Followed trip has no destination")
        return Itinerary(itineraryOrigin, itineraryDestination, tripLegs.map { it.toTripLeg() })
    }

    companion object {
        const val CURRENT_VERSION = 2

        @JvmStatic
        fun fromItinerary(itinerary: Itinerary): FollowedTripRecord =
            FollowedTripRecord().apply {
                origin = itinerary.origin.abbreviation
                trainDestination = itinerary.trainDestination?.abbreviation
                passengerDestination = itinerary.destination.abbreviation
                line = itinerary.line?.name
                platform = itinerary.platform
                canceled = itinerary.canceled
                minEstimate = itinerary.getInitialDepartureTime()
                maxEstimate = itinerary.getInitialDepartureTime()
                estimatedTripTime = (
                    itinerary.getEstimatedArrivalTime() - itinerary.getInitialDepartureTime()
                ).coerceAtLeast(0L).toInt()
                tripLegs = itinerary.legs.map(TripLegRecord::fromTripLeg).toMutableList()
            }

        private fun abbreviation(station: Station?): String? = station?.abbreviation

        private fun station(abbreviation: String?): Station? =
            abbreviation?.let(Station::getByAbbreviation)
    }

    class TripLegRecord {
        @JvmField var line: String? = null
        @JvmField var origin: String? = null
        @JvmField var destination: String? = null
        @JvmField var trainDestination: String? = null
        @JvmField var tripId: String? = null
        @JvmField var departureTime = 0L
        @JvmField var arrivalTime = 0L
        @JvmField var minimumTransferSecondsAfter = 0
        @JvmField var scheduledDepartureTime = 0L
        @JvmField var scheduledArrivalTime = 0L
        @JvmField var departureSource: String? = null
        @JvmField var arrivalSource: String? = null
        @JvmField var platform: String? = null
        @JvmField var serviceDate: String? = null
        @JvmField var canceled = false
        @JvmField var direction: String? = null
        @JvmField var stops: MutableList<TripStopRecord> = mutableListOf()

        fun toTripLeg(): TripLeg = TripLeg(
            line?.let(Line::valueOf),
            station(origin),
            station(destination),
            station(trainDestination),
            tripId,
            departureTime,
            arrivalTime,
            stops.map { it.toTripStop() },
            minimumTransferSecondsAfter,
            scheduledDepartureTime,
            scheduledArrivalTime,
            predictionSource(departureSource),
            predictionSource(arrivalSource),
            platform,
            serviceDate?.let { value -> runCatching { LocalDate.parse(value) }.getOrNull() },
            canceled,
            direction,
        )

        companion object {
            @JvmStatic
            fun fromTripLeg(leg: TripLeg): TripLegRecord = TripLegRecord().apply {
                line = leg.line?.name
                origin = abbreviation(leg.origin)
                destination = abbreviation(leg.destination)
                trainDestination = abbreviation(leg.trainDestination)
                tripId = leg.tripId
                departureTime = leg.departureTime
                arrivalTime = leg.arrivalTime
                minimumTransferSecondsAfter = leg.minimumTransferSecondsAfter
                scheduledDepartureTime = leg.scheduledDepartureTime
                scheduledArrivalTime = leg.scheduledArrivalTime
                departureSource = leg.departureSource.name
                arrivalSource = leg.arrivalSource.name
                platform = leg.platform
                serviceDate = leg.serviceDate?.toString()
                canceled = leg.canceled
                direction = leg.direction
                stops = leg.stops.map(TripStopRecord::fromTripStop).toMutableList()
            }
        }
    }

    class TripStopRecord {
        @JvmField var station: String? = null
        @JvmField var arrivalTime = 0L
        @JvmField var departureTime = 0L
        @JvmField var scheduledArrivalTime = 0L
        @JvmField var scheduledDepartureTime = 0L
        @JvmField var arrivalSource: String? = null
        @JvmField var departureSource: String? = null

        fun toTripStop(): TripStop = TripStop(
            station(station),
            arrivalTime,
            departureTime,
            scheduledArrivalTime,
            scheduledDepartureTime,
            predictionSource(arrivalSource),
            predictionSource(departureSource),
        )

        companion object {
            @JvmStatic
            fun fromTripStop(stop: TripStop): TripStopRecord = TripStopRecord().apply {
                station = stop.station?.abbreviation
                arrivalTime = stop.arrivalTime
                departureTime = stop.departureTime
                scheduledArrivalTime = stop.scheduledArrivalTime
                scheduledDepartureTime = stop.scheduledDepartureTime
                arrivalSource = stop.arrivalSource.name
                departureSource = stop.departureSource.name
            }
        }
    }
}

private fun predictionSource(value: String?): PredictionSource =
    value?.let { runCatching { PredictionSource.valueOf(it) }.getOrNull() }
        ?: PredictionSource.UNKNOWN

private fun station(abbreviation: String?): Station? =
    abbreviation?.let(Station::getByAbbreviation)
